client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java

```diff
@@ -1699,16 +1699,25 @@ public CelebornInputStream readPartition(
       int startMapIndex,
       int endMapIndex,
       ExceptionMaker exceptionMaker,
+      ArrayList<PartitionLocation> locations,
+      ArrayList<PbStreamHandler> streamHandlers,
+      int[] mapAttempts,
       MetricsCallback metricsCallback)
       throws IOException {
     if (partitionId == Utils$.MODULE$.UNKNOWN_APP_SHUFFLE_ID()) {
       logger.warn("Shuffle data is empty for shuffle {}: UNKNOWN_APP_SHUFFLE_ID.", shuffleId);
       return CelebornInputStream.empty();
     }
-    ReduceFileGroups fileGroups = updateFileGroup(shuffleId, partitionId);
 
-    if (fileGroups.partitionGroups.isEmpty()
-        || !fileGroups.partitionGroups.containsKey(partitionId)) {
+    if (mapAttempts == null) {
+      ReduceFileGroups fileGroups = updateFileGroup(shuffleId, partitionId);
+      mapAttempts = fileGroups.mapAttempts;
```

Comment:
why not use locations as the condition to refresh fileGroups? It's more clear then mapAttempts.

Reply:
That's because `locations` can be empty even though BatchOpenStream has been tried. If `mapAttempts` is null, it's guaranteed that BatchOpenStream has not been tried.

Reply:
Commented, PTAL @RexXiong